src/hyperlink/_url.py

```diff
@@ -1705,86 +1821,111 @@ def child(self, *segments):
         new_segs = [_encode_reserved(s) for s in segments]
         return self.__class__(self._url.child(*new_segs))
 
-    def normalize(self, *a, **kw):
+    def normalize(self, scheme=True, host=True, path=True, query=True,
+                  fragment=True, userinfo=True, percents=True):
+        # type: (bool, bool, bool, bool, bool, bool, bool) -> DecodedURL
         """Return a new `DecodedURL` wrapping the result of
         :meth:`~hyperlink.URL.normalize()`
         """
-        return self.__class__(self._url.normalize(*a, **kw))
+        return self.__class__(self._url.normalize(
+            scheme, host, path, query, fragment, userinfo, percents
+        ))
 
     @property
     def absolute(self):
+        # type: () -> bool
         return self._url.absolute
 
     @property
     def scheme(self):
+        # type: () -> Text
         return self._url.scheme
 
     @property
     def host(self):
+        # type: () -> Text
         return _decode_host(self._url.host)
 
     @property
     def port(self):
+        # type: () -> Optional[int]
         return self._url.port
 
     @property
     def rooted(self):
+        # type: () -> bool
         return self._url.rooted
 
     @property
     def path(self):
-        try:
-            return self._path
-        except AttributeError:
-            pass
-        self._path = tuple([_percent_decode(p, raise_subencoding_exc=True)
-                            for p in self._url.path])
+        # type: () -> Sequence[Text]
+        if not hasattr(self, "_path"):
+            self._path = tuple([
+                _percent_decode(p, raise_subencoding_exc=True)
+                for p in self._url.path
+            ])
```

Comment:
Checking for the attribute instead of catching `AttributeError` helps `mypy` out.
I think it's a little more readable as well… (but only a little).

Reply:
Weird that mypy has trouble with that, given Python's classic argument about TOCTOU / EAFP. 

Another issue is that `hasattr()` returns False on non-AttributeErrors on py2. But realistically it's not a big regression, so it's not a blocker.